### DISCUSSION TITLE
Simple audio mixer implementation

### DISCUSSION THREAD

**Discussion by leo150** [HaishinKit/HaishinKit.swift#1373] [Idea]
Hey. First of all, thanks a lot for adding all the audio enhancements in the past several months. They help a lot with the audio experience, especially handling various sample rates.

Regarding further audio features, would it be possible to add a simple audio mixer at some point? This would greatly help with improving screen sharing functionality. This way, users can stream their mic and ReplayKit audio at the same time. I'm currently using a mixer based on the [VideoCast](https://github.com/cats-oss/VideoCast-Swift/blob/master/Source/Mixers/AudioMixer.swift) implementation. There are also a couple of open-source implementations, such as [PLAudioMixer](https://github.com/PatrickSCLin/PLAudioMixer/blob/master/PLAudioMixer/PLAudioMixer.swift), [VideoCore](https://github.com/jgh-/VideoCore-Inactive/blob/master/mixers/GenericAudioMixer.cpp), [SwiftVideo](https://github.com/unpause-live/SwiftVideo/blob/master/Sources/SwiftVideo/mix.audio.swift).

I can create a PR with one of the implementations, but there are architectural decisions I'm not sure about in the context of HaishinKit. First, audio formats need to be converted to match each other before mixing. HK already does this, so the mixer needs to be placed in a way that avoids doing the conversion twice. Second, HK is using queues; I'm not sure if some of them can be reused as well. Any advice will be appreciated. Thanks.

**Comment by shogo4405** [comment]
You're welcome. Regarding audio mixing, I'm considering implementing it since macOS provides multiple resources. However, I'd like to first add a lightweight mixing feature for videos. Therefore, I anticipate that the implementation might take some time.

When utilizing other open-source libraries, it's advisable to maintain loose coupling by inputting the CMSampleBuffer or AVAudioBuffer produced by the library into HaishinKit.
```
CMSampleBuffer (ReplayKit audio). ──┐
	              Mixing Audio Library ─> CMSampleBuffer or AVAudioBuffer ─> HaishinKit(append(_ sampleBuffer: CMSampleBuffer) or append(audioBuffer: AVAudioBuffer))
CMSampleBuffer (ReplayKit mic) ───┛
```

Continuing processing on the HaishinKit side involves adjusting the sample count and preserving silent parts. I consider this a necessary process and a cost that should be paid.

**Comment by dumojo** [comment]
> I can create a PR with one of the implementations, but there are architectural decisions I'm not sure about in the context of HaishinKit. First, audio formats need to be converted to match each other before mixing. HK already does this, so the mixer needs to be placed in a way that avoids doing the conversion twice. Second, HK is using queues; I'm not sure if some of them can be reused as well. Any advice will be appreciated. Thanks.

Hi if you have a properly working audio mixer implementation for CMSampleBuffer/AVAudioPCMBuffer it would be great if you could share it with me.

**Reply by leo150** [reply]
https://github.com/shogo4405/HaishinKit.swift/pull/1386

**Comment by shogo4405** [comment]
Considering ReplayKit, I feel there is a volume difference between appAudio and micAudio. Specifically, appAudio seems too loud. Is there an easy way to control the volume? https://github.com/shogo4405/HaishinKit.swift/pull/1386

**Reply by leo150** [reply]
Yes here's a stub method [here](https://github.com/shogo4405/HaishinKit.swift/blob/c927f04b445b1ca0c146a85242df223e3f120a8b/Sources/IO/AudioNode.swift#L359) for the future implementation. I don't recall testing it, but it should work right away.

**Reply by shogo4405** [reply]
I was able to confirm that the volume can be changed using the method you provided. Thank you.
https://github.com/shogo4405/HaishinKit.swift/blob/main/Sources/IO/IOAudioMixerByMultiTrack.swift#L13-L15
